lib/ansible/modules/cloud/smartos/vmadm.py

```diff
@@ -0,0 +1,609 @@
+        match = re.match('Successfully created VM (.*)', stderr)
+        if match:
+            vm_uuid = match.groups()[0]
+            if not is_valid_uuid(vm_uuid):
+                module.fail_json(msg='Invalid UUID for VM {0}?'.format(vm_uuid))
+        else:
+            module.fail_json(msg='Could not retrieve UUID of newly created(?) VM')
+
+        # Now that the VM is created, ensure it is in the desired state (if not 'running')
+        if vm_state != 'running':
+            ret = set_vm_state(module, vm_uuid, vm_state)
+            if not ret:
+                module.fail_json(msg='Could not set VM {0} to state {1}'.format(vm_uuid, vm_state))
+
+    try:
+        os.unlink(payload_file)
+    except Exception as e:
+        # Since the payload may contain sensitive information, fail hard
+        # if we cannot remove the file so the operator knows about it.
+        module.fail_json(msg='Could not remove temporary JSON payload file {0}: {1}'.
+                         format(payload_file, e))
+
+    return changed, vm_uuid
+
+
+def vmadm_create_vm(module, payload_file):
+    """Create a new VM using the provided payload."""
+    cmd = '{0} create -f {1}'.format(module.vmadm, payload_file)
+
+    return module.run_command(cmd)
+
+
+def set_vm_state(module, vm_uuid, vm_state):
+    p = module.params
+
+    # Check if the VM is already in the desired state.
+    state = get_vm_prop(module, vm_uuid, 'state')
+    if state and (state == vm_state):
+        return None
+
+    # Lookup table for the state to be in, and which command to use for that.
+    # vm_state: [vmadm commandm, forceable?]
+    cmds = {
+        'stopped': ['stop', True],
+        'running': ['start', False],
+        'deleted': ['delete', True],
+        'rebooted': ['reboot', False]
+    }
+
+    if p['force'] and cmds[vm_state][1]:
+        force = '-F'
+    else:
+        force = ''
+
+    cmd = 'vmadm {0} {1} {2}'.format(cmds[vm_state][0], force, vm_uuid)
+
+    (rc, stdout, stderr) = module.run_command(cmd)
+
+    match = re.match('^Successfully.*', stderr)
+    if match:
+        return True
+    else:
+        return False
+
+
+def create_payload(module, uuid):
+    """Create the JSON payload (vmdef) and return the filename."""
+
+    p = module.params
+
+    # Filter out the few options that are not valid VM properties.
+    module_options = ['debug', 'force', 'state']
+    vmattrs = filter(lambda prop: prop not in module_options, p)
+
+    vmdef = {}
+
+    for attr in vmattrs:
+        if p[attr]:
+            vmdef[attr] = p[attr]
+
+    try:
+        vmdef_json = json.dumps(vmdef)
+    except Exception as e:
+        module.fail_json(msg='Could not create valid JSON payload: {0}'.format(e))
+
+    # Create the temporary file that contains our payload, and set tight
+    # permissions for it may container sensitive information.
+    try:
+        # XXX: When there's a way to get the current ansible temporary directory
```

Comment:
Well, as mentioned by bcoca, there should be a generic function that will tell us the name of the temporary directory created by Ansible. Once we have said function it can be used here.